// @n8n/expression-runtime/dist/cjs/observability/metrics.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.EXPRESSION_METRICS = void 0;
    exports.EXPRESSION_METRICS = {
        evaluationDuration: {
            name: 'expression.evaluation.duration_seconds',
            kind: 'histogram',
            labels: ['status', 'type'],
            help: 'Duration of VM-based expression evaluation in seconds.',
        },
        codeCacheHit: {
            name: 'expression.code_cache.hit',
            kind: 'counter',
            labels: [],
            help: 'Expression code cache hits.',
        },
        codeCacheMiss: {
            name: 'expression.code_cache.miss',
            kind: 'counter',
            labels: [],
            help: 'Expression code cache misses.',
        },
        codeCacheEviction: {
            name: 'expression.code_cache.eviction',
            kind: 'counter',
            labels: [],
            help: 'Expression code cache evictions.',
        },
        codeCacheSize: {
            name: 'expression.code_cache.size',
            kind: 'gauge',
            labels: [],
            help: 'Current expression code cache size.',
        },
        poolAcquired: {
            name: 'expression.pool.acquired',
            kind: 'counter',
            labels: [],
            help: 'Bridges acquired from the expression pool.',
        },
        poolReplenishFailed: {
            name: 'expression.pool.replenish_failed',
            kind: 'counter',
            labels: [],
            help: 'Failed pool bridge replenishments.',
        },
        poolScaledUp: {
            name: 'expression.pool.scaled_up',
            kind: 'counter',
            labels: [],
            help: 'Expression pool scaled up from idle.',
        },
        poolScaledToZero: {
            name: 'expression.pool.scaled_to_zero',
            kind: 'counter',
            labels: [],
            help: 'Expression pool scaled to zero after inactivity.',
        },
    };
});
